src/Security/SecProtocolMetadata.cs

```diff
@@ -51,9 +51,181 @@ public class SecProtocolMetadata : NativeObject {
 
 		public bool EarlyDataAccepted => sec_protocol_metadata_get_early_data_accepted (GetCheckedHandle ()) != 0;
 
-		//
-		// MISSING: all the block APIs
-		//
+ 		[DllImport (Constants.SecurityLibrary)]
+		[return: MarshalAs (UnmanagedType.I1)]
+		extern static bool sec_protocol_metadata_challenge_parameters_are_equal (IntPtr metadataA, IntPtr metadataB);
 
+		public static bool ChallengeParametersAreEqual (SecProtocolMetadata metadataA, SecProtocolMetadata metadataB)
+		{
+			if (metadataA == null)
+				throw new ArgumentNullException (nameof (metadataA));
+			if (metadataB == null)
+				throw new ArgumentNullException (nameof (metadataB));
+			return sec_protocol_metadata_challenge_parameters_are_equal (metadataA.GetCheckedHandle (), metadataB.GetCheckedHandle ());
+		}
+
+ 		[DllImport (Constants.SecurityLibrary)]
+		[return: MarshalAs (UnmanagedType.I1)]
+		extern static bool sec_protocol_metadata_peers_are_equal (IntPtr metadataA, IntPtr metadataB);
+
+		public static bool PeersAreEqual (SecProtocolMetadata metadataA, SecProtocolMetadata metadataB)
+		{
+			if (metadataA == null)
+				throw new ArgumentNullException (nameof (metadataA));
+			if (metadataB == null)
+				throw new ArgumentNullException (nameof (metadataB));
+			return sec_protocol_metadata_peers_are_equal (metadataA.GetCheckedHandle (), metadataB.GetCheckedHandle ());
+		}
+
+#if !COREBUILD
+
+		delegate void sec_protocol_metadata_access_distinguished_names_handler_t (IntPtr block, IntPtr dispatchData);
+ 		static sec_protocol_metadata_access_distinguished_names_handler_t static_DistinguishedNamesForPeer  = TrampolineDistinguishedNamesForPeer;
+
+ 		[MonoPInvokeCallback (typeof (sec_protocol_metadata_access_distinguished_names_handler_t))]
+ 		static void TrampolineDistinguishedNamesForPeer (IntPtr block, IntPtr data)
+ 		{
+ 			var del = BlockLiteral.GetTarget<Action<DispatchData>> (block);
+ 			if (del != null) {
+ 				var dispatchData = new DispatchData (data, owns: false);
+ 				del (dispatchData);
+ 			}
+ 		}
+
+ 		[DllImport (Constants.SecurityLibrary)]
+ 		static extern unsafe void sec_protocol_metadata_access_distinguished_names (IntPtr handle, void *callback);
+
+ 		[BindingImpl (BindingImplOptions.Optimizable)]
+ 		public void SetDistinguishedNamesForPeerHandler (Action<DispatchData> callback)
+ 		{
+ 			unsafe {
+ 				if (callback == null) {
```

Comment:
The native function is an iterator, which means that a null callback isn't valid.

This means a few things:

1. The managed function name could be better. `AccessDistinguishedNames` is what would most match the native name, but someone else might have a better suggestion.
2. Throw ArgumentNullException if callback is null.
3. The P/Invoke can use `ref BlockLiteral` (and not be unsafe):

```csharp
static extern void sec_protocol_metadata_access_distinguished_names (IntPtr handle, ref BlockLiteral callback);
```

4. The method that calls the P/Invoke doesn't have to be unsafe either:

```csharp
BlockLiteral block_handler = new BlockLiteral ();
block_handler.SetupBlockUnsafe (static_DistinguishedNamesForPeer, callback);

try {
	sec_protocol_metadata_access_distinguished_names (GetCheckedHandle (), ref block_handler);
} finally {
	block_handler.CleanupBlock ();
}
```

5. I wonder if we should also add an overload that does not take a block, but returns an array of all the iterated elements (@spouliot?):

```csharp
public IList<DispatchData> GetDistinguishedNames ()
{
    var rv = new List<DispatchData> ();
    AccessDistinguishedNames ((v) => rv.Add (v));
    return rv;
}
```

It seems a bit overkill to have an iterator function when there's no way to stop iterating in the middle.

Reply:
Thanks for this! I'll update the iterator ones. I personally like the idea of the overload, but lets wait for sebs input. I'll fix the other comments and will push. We can add the overload later.